xarray/tests/test_dataarray.py

```diff
@@ -4571,6 +4571,48 @@ def sine(t, a, f, p):
                 bounds={"a": (0, DataArray([1], coords={"foo": [1]}))},
             )
 
+    @requires_scipy
+    @pytest.mark.parametrize("use_dask", [True, False])
+    def test_curvefit_allow_failures(self, use_dask: bool) -> None:
+        if use_dask and not has_dask:
+            pytest.skip("requires dask")
+
+        # nonsense function to make the optimization fail
+        def line(x, a, b):
+            if a > 10:
+                return 0
+            return a * x + b
+
+        da = DataArray(
+            [[1, 3, 5], [0, 20, 40]],
+            coords={"i": [1, 2], "x": [0.0, 1.0, 2.0]},
+        )
+
+        if use_dask:
+            da = da.chunk({"i": 1})
+
+        expected = DataArray(
+            [[2, 1], [np.nan, np.nan]], coords={"i": [1, 2], "param": ["a", "b"]}
+        )
+
+        with pytest.raises(RuntimeError):
+            da.curvefit(
```

Comment:
Is the scipy curvefit deterministic?
Maybe better to add a p0 to ensure that it always fails and not randomly converges in 5 iterations.

Reply:
Good point. I think it is deterministic, I would expect this to only break if scipy changes what it's doing. But in this case the default p0 is actually calculated by xarray, in `_initialize_curvefit_params`. But I can also add an explicit p0.

Reply:
Woops, looks like it's indeed not good enough, the test failed in the CI even though it worked for me. I'll try to come up with a better test but I might not have time to do it today.

Reply:
Aha, the problem was that I forgot to call `compute()`, which then didn't produce the expected error in the dask version (I forgot to test that locally). I fixed this in f61cb4f4. I also wrote a version with `maxfev=1`, and an explicit initial guess, but to get it to work  I had to make the target function a bit more weird, see here: https://github.com/mgunyho/xarray/commit/b46c804ccad73bfb5bbf4d80c9889f780b41e481. Do you think this is better?

Reply:
Yes, looks good.
No idea if that solves the problem though.

Reply:
Just to double-check: you mean that the maxfev=1 version in https://github.com/mgunyho/xarray/commit/b46c804ccad73bfb5bbf4d80c9889f780b41e481 is good? I can add it to this PR if so. It does solve the problem, but it was already solved by just adding `compute()` in https://github.com/pydata/xarray/commit/f61cb4f45052e6ad5a98ec336b87223246d9f2b7, now the CI only failed because of the flaky test which is unrelated to this PR.

Reply:
Ah sorry I misunderstood.
Yes it is fine now!